### PR TITLE
Melsumner/issue 378

### PR DESCRIPTION
**Pull request by MelSumner**
If merged, this PR resolves the issue described in #378 

## Details

- added 3 additional tests for potential combinations
- updated the `hasQueryParams` check so that it checks for more combinations
- updated the `hasQueryParams` check to be more terse
- updated the description of the getter/param
- added a check in the `onRouteChange` action to compare routes

After review/any changes/approval, I'll squash/merge and do a patch release.

### CODE REVIEW

**Comment by MelSumner** [comment]
@deanmarano can you check this and see if this is what you would expect? TYSM!

**Review by ef4 — APPROVED** [review]
This change makes sense to me.

File: tests/integration/components/navigation-narrator-test.js [hunk 1]
```diff
@@ -173,5 +173,77 @@ module('Integration | Component | navigation-narrator', function (hooks) {
...
+          to: new MockRouteInfo({
+            name: 'biscuit',
+            params: { id: 'digestive' },
+          }),
+        })
+      );
+
+      await settled();
+
+      assert.dom('#ember-a11y-refocus-nav-message').isNotFocused();
```
Comment (MelSumner) [inline tests/integration/components/navigation-narrator-test.js:246]: I just realized that this test isn't correct. Will need to review more ><

File: tests/integration/components/navigation-narrator-test.js [hunk 2]
```diff
@@ -173,5 +173,77 @@ module('Integration | Component | navigation-narrator', function (hooks) {
...
+    });
+
+    test('excludeAllQueryParams is true, queryParams do exist on the route but then are emptied, still do not manage focus', async function (assert) {
```
Comment (MelSumner) [inline tests/integration/components/navigation-narrator-test.js:201]: This test is meant to cover a scenario like this:

`myurl.com/someRoute?someQueryParam=foo` -->  `myurl.com/someRoute`

File: tests/integration/components/navigation-narrator-test.js [hunk 3]
```diff
@@ -173,5 +173,77 @@ module('Integration | Component | navigation-narrator', function (hooks) {
...
+          }),
+        })
+      );
+
+      await settled();
+
+      assert.dom('#ember-a11y-refocus-nav-message').isNotFocused();
+    });
+
+    test('excludeAllQueryParams is true, queryParams do exist on the route, user navigates to new route, manage focus', async function (assert) {
```
Comment (MelSumner) [inline tests/integration/components/navigation-narrator-test.js:225]: This test is meant to cover a scenario like this:

`myurl.com/someRoute?someQueryParam=foo` -->  `myurl.com/someOtherRoute`

File: tests/integration/components/navigation-narrator-test.js [hunk 4]
```diff
@@ -173,5 +173,77 @@ module('Integration | Component | navigation-narrator', function (hooks) {
 
       assert.dom('#ember-a11y-refocus-nav-message').isFocused();
     });
+
+    test('excludeAllQueryParams is true, queryParams do not exist but are added, do not manage focus', async function (assert) {
```
Comment (MelSumner) [inline tests/integration/components/navigation-narrator-test.js:177]: This test is meant to cover a scenario like this:

`myurl.com/someRoute` -->  `myurl.com/someRoute?someQueryParam=foo`

File: tests/integration/components/navigation-narrator-test.js [hunk 5]
```diff
@@ -173,5 +173,88 @@ module('Integration | Component | navigation-narrator', function (hooks) {
...
+          }),
+        })
+      );
+
+      await settled();
+
+      assert.dom('#ember-a11y-refocus-nav-message').isNotFocused();
+    });
+
+    test('excludeAllQueryParams is true, queryParams exist on the route but then are emptied, still do not manage focus', async function (assert) {
```
Comment (MelSumner) [inline tests/integration/components/navigation-narrator-test.js:201]: This test is meant to cover a scenario like this:

`myurl.com/someRoute?someQueryParam=foo` -->  `myurl.com/someRoute`

File: tests/integration/components/navigation-narrator-test.js [hunk 6]
```diff
@@ -173,5 +173,88 @@ module('Integration | Component | navigation-narrator', function (hooks) {
...
+          }),
+        })
+      );
+
+      await settled();
+
+      assert.dom('#ember-a11y-refocus-nav-message').isNotFocused();
+    });
+
+    test('excludeAllQueryParams is true, queryParams exist on the route, user navigates to new route without any query params, manage focus', async function (assert) {
```
Comment (MelSumner) [inline tests/integration/components/navigation-narrator-test.js:225]: This test is meant to cover a scenario like this:

`myurl.com/someRoute?someQueryParam=foo` -->  `myurl.com/someOtherRoute`

This test is currently failing, though. @jgwhite could you take a look and just confirm if the `MockRouteInfo` syntax is correct here (or not)? I inferred this from the test helper. 

If it is, I might need to re-visit the solution.

File: addon/components/navigation-narrator.js [hunk 7]
```diff
@@ -108,7 +107,9 @@ export default class NavigationNarratorComponent extends Component {
     let shouldFocus;
     this.transition = transition; // We need to do this because we can't pass an argument to a getter
 
-    if (this.excludeAllQueryParams && this.hasQueryParams) {
+    let hasSameRoute = this.transition.from.name === this.transition.to.name;
+
+    if (this.excludeAllQueryParams && this.hasQueryParams && hasSameRoute) {
```
Comment (MelSumner) [inline addon/components/navigation-narrator.js:112]: This makes the failing test pass, but I also wonder if there's a more elegant way to do this. Suggestions welcome.

File: tests/integration/components/navigation-narrator-test.js [hunk 8]
```diff
@@ -173,5 +173,88 @@ module('Integration | Component | navigation-narrator', function (hooks) {
...
+      let router = this.owner.lookup('service:router');
+
+      await render(hbs`<NavigationNarrator @excludeAllQueryParams={{true}} />`);
+      router.trigger(
+        'routeDidChange',
+        new MockTransition({
```
Comment (MelSumner) [inline tests/integration/components/navigation-narrator-test.js:231]: I used the same format here: https://github.com/ember-a11y/ember-a11y-refocus/blob/7d20ca2f5d82b24d13d550257d4a61d0e64de7f9/tests/unit/utils/default-validator-test.js#L35

File: addon/components/navigation-narrator.js (hunk 7 above)
Comment (deanmarano) [reply addon/components/navigation-narrator.js:112]: The `this.hasQueryParams` might not be needed with the `hasSameRoute`.

**Pull request merged**